### PR TITLE
Node 12 compatibility

### PR DESCRIPTION
**Pull request by Algunenano**
This PR does a lot of changes to update to a new NaN with node 12 compatibility. Not all changes were mandatory, as most address thing that are currently deprecated (warnings) but not removed.

The commits can (and probably should) be reviewed individually.

The most controversial is the last one (d0bff89): It appears that in node an empty vector tile `toJSON()` returns `[]` instead of `{}`, which makes more sense since it returns an empty array but I'm not sure exactly what change in node12 has triggered this change in behaviour and whether you might want to add some kind of warning in the documentation.

References https://github.com/mapnik/node-mapnik/issues/921

### CODE REVIEW

**Comment by flippmoke** [comment]
I am working through this PR still, I am sorry its just massive and changes lots of things (for the better) so it might take a bit.

**Comment by Algunenano** [comment]
No worries. Thanks a lot for the effort

**Comment by flippmoke** [comment]
Not at all complete in my review, but ran into thoughts as I thought about how we handle Async. In other libraries we have shifted to using a design like this: 

https://github.com/mapbox/node-cpp-skel/blob/master/src/standalone_async/hello_async.cpp#L73

Should we shift to doing something like this and kill the "baton" concept for good in a change like this? Using `Nan::AsyncResource async_resource(__func__);` seems a little suspect to me.

File: src/blend.cpp [hunk 1]
```diff
@@ -89,55 +89,55 @@ NAN_METHOD(hsl2rgb) {
...
+    h = Nan::To<double>(info[0]).FromJust();
+    s = Nan::To<double>(info[1]).FromJust();
+    l = Nan::To<double>(info[2]).FromJust();
     v8::Local<v8::Array> rgb = Nan::New<v8::Array>(3);
     std::uint32_t r,g,b;
     hsl_to_rgb(h,s,l,r,g,b);
-    rgb->Set(0,Nan::New<v8::Integer>(r));
-    rgb->Set(1,Nan::New<v8::Integer>(g));
-    rgb->Set(2,Nan::New<v8::Integer>(b));
+    Nan::Set(rgb,  0,Nan::New<v8::Integer>(r));
```
[the remark this answers is not in the document]
Comment (Algunenano) [reply src/blend.cpp]: I'm addressing it. Have you considered adding a `.clang-format` file or something similar to check the style. It'd make my life easier as I could set up a check with a precommit hook.

**Comment by Algunenano** [comment]
> https://github.com/mapbox/node-cpp-skel/blob/master/src/standalone_async/hello_async.cpp#L73

> Should we shift to doing something like this and kill the "baton" concept for good in a change like this? 

Yes, it'd nice to move away from using libuv directly. One thing to note for that example is that I don't see how to handle references to other native javascript objects that you need to keep alive until the callback is done. Do you add them to the class and a reference in the constructor and a release in the constructor or am I missing something and it's handled automagically?

Anyhow, I went for `Nan::AsyncResource` because it was what required the least changes / effort in the current setup, so I'd prefer having a different PR to change all the baton concept into `Nan::AsyncWorker`.

> Using Nan::AsyncResource async_resource(__func__); seems a little suspect to me.

If you are refering to the `__func__` name I chose because it makes for an unique name and easy to find the function if it gets logged somewhere (I don't know what the async hook type is really used for).

File: src/mapnik_feature.cpp [hunk 2]
```diff
@@ -114,7 +114,7 @@ NAN_METHOD(Feature::fromJSON)
         }
         Feature* feat = new Feature(f);
         v8::Local<v8::Value> ext = Nan::New<v8::External>(feat);
-        Nan::MaybeLocal<v8::Object> maybe_local = Nan::NewInstance(Nan::New(constructor)->GetFunction(), 1, &ext);
+        v8::MaybeLocal<v8::Object> maybe_local = Nan::NewInstance(Nan::GetFunction(Nan::New(constructor)).ToLocalChecked(), 1, &ext);
```
Comment (flippmoke) [inline src/mapnik_feature.cpp]: The return type from ` Nan::NewInstance` is `Nan::MaybeLocal<v8::Object>` I think, why the shift to `v8::MaybeLocal<v8::Object>`?

Comment (Algunenano) [reply src/mapnik_feature.cpp]: Seems like an oversight (maybe I copied it from somewhere else in the code?). I'll fix those places where I've modified to use Nan::MaybeLocal instead.

**Comment by artemp** [comment]
Great, thanks ^ 
Should we update `mapnik-vector-tile' `3.0.0` ? Also, I see a few complier warnings with clang-8 
/cc @Algunenano @flippmoke

**Comment by Algunenano** [comment]
> Should we update mapnik-vector-tile 3.0.0 ? 

I'm personally not using MVTs through node-mapnik anymore, but I'd say yes.

> Also, I see a few complier warnings with clang-8

I also get some  warnings, one unused variable (src/mapnik_vector_tile.cpp:1366:27) and also some coming from `mapnik-vector-tile` and `boost/gil` that should be addressed in their own repos.

**Pull request merged**